Use inventory reservoir as source for all files and dirs

arunravimv opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://github.com/delta-io/delta/blob/master/CONTRIBUTING.md
  2. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP] Your PR title ...'.
  3. Be sure to keep the PR description updated to reflect all changes.
  4. Please write your PR title to summarize what this PR proposes.
  5. If possible, provide a concise example to reproduce the issue for a faster review.
  6. If applicable, include the corresponding issue number in the PR title and link it in the body.
-->

#### Which Delta project/connector is this regarding?
<!--
Please add the component selected below to the beginning of the pull request title
For example: [Spark] Title of my pull request
-->

- [x] Spark
- [ ] Standalone
- [ ] Flink
- [ ] Kernel
- [ ] Other (fill in here)

## Description
- Currently, users have large tables with daily/hourly partitions for many years, among all these partitions only recent ones are subjected to change due to job reruns, corrections, and late arriving events.

- When Vacuum is run on these tables, the listing of files is performed on all the partitions and it runs for several hours/days. This duration grows as tables grow and vacuum becomes a major overhead for customers especially when they have hundreds or thousands of such delta tables. File system scan takes the most amount of time in Vacuum operation for large tables, mostly due to the parallelism achievable and API throttling on the object stores.

- This change provides a way for users to pass a reservoir of files generated externally (eg: from inventory reports of cloud stores) as a delta table or as a spark SQL query (having a predefined schema). The vacuum operation when provided with such a reservoir data frame will skip the listing operation and use it as a source of all files in the storage.

"Resolves #1691".

## How was this patch tested?

- Unit Testing  (` build/sbt 'testOnly org.apache.spark.sql.delta.DeltaVacuumSuite'`)

## Does this PR introduce _any_ user-facing changes?

yes, the MR accepts an optional method to pass inventory.

`VACUUM table_name [USING INVENTORY <reservoir-delta-table>] [RETAIN num HOURS] [DRY RUN]`
`VACUUM table_name [USING INVENTORY <reservoir-query>] [RETAIN num HOURS] [DRY RUN]`

eg:  `VACUUM test_db.table using inventory select * from reservoir_table RETAIN 168 HOURS dry run`

bart-samwel commented on spark/src/main/scala/io/delta/sql/parser/DeltaSqlParser.scala:
```diff
@@ -310,6 +310,9 @@ class DeltaSqlAstBuilder extends DeltaSqlBaseBaseVisitor[AnyRef] {
     VacuumTableCommand(
       Option(ctx.path).map(string),
       Option(ctx.table).map(visitTableIdentifier),
+      Option(ctx.reservoirPath).map(string),
```
Can we use named arguments here? Because we have a number of same-type parameters here it's very easy to get the order wrong.

bart-samwel commented on spark/src/main/scala/io/delta/tables/execution/VacuumTableCommand.scala:
```diff
@@ -36,6 +37,8 @@ import org.apache.spark.sql.types.StringType
 case class VacuumTableCommand(
     override val child: LogicalPlan,
     horizonHours: Option[Double],
+    reservoirTable: Option[LogicalPlan],
```
The external API is called "inventory" but internally it's called "reservoir". IIUIC the term "reservoir" is used in this code but it's meant to be the set of files that are _actually_ in the reservoir (e.g. `reservoirBase` is the base path of the reservoir). Would it make sense to call this "inventory", to match the syntax? And then the "reservoir files" are given by the intersection of the inventory with the reservoirBase.

arunravimv replied:
@bart-samwel :) I was a bit skeptical about using inventory as it was too AWS-specific (rightly pointed out by you in the other discussion thread). So reservoir was the other name that popped into my mind. If you are OK with the reservoir I will update the grammar to use the reservoir instead of the AWS specific inventory term. If you think we can go with Inventory ;) then I will update the code and functions.

bart-samwel commented on spark/src/main/scala/io/delta/tables/execution/VacuumTableCommand.scala:
```diff
@@ -53,17 +56,33 @@ case class VacuumTableCommand(
         DeltaTableIdentifier(path = Some(deltaTable.path.toString)))
     }
-    VacuumCommand.gc(sparkSession, deltaTable.deltaLog, dryRun, horizonHours).collect()
+    val reservoir = reservoirTable match {
+      case Some(table) => Some(
+        getDeltaTable(sparkSession.sessionState.analyzer.execute(table), "VACUUM")
+          .toDf(sparkSession)
+      )
+      case None => reservoirQuery.map(sparkSession.sql)
+    }
```
Equivalent but shorter?
```suggestion
    val reservoir = reservoirTable.map(table =>
        getDeltaTable(sparkSession.sessionState.analyzer.execute(table), "VACUUM")
          .toDf(sparkSession)
      ).getOrElse(reservoirQuery.map(sparkSession.sql))
```

bart-samwel commented on spark/src/main/antlr4/io/delta/sql/parser/DeltaSqlBase.g4:
```diff
@@ -73,6 +73,7 @@ singleStatement
 // If you add keywords here that should not be reserved, add them to 'nonReserved' list.
 statement
     : VACUUM (path=STRING | table=qualifiedName)
+        (USING INVENTORY (reservoirPath=STRING | reservoirTable=qualifiedName |  reservoirQuery=statement))?
```
Can we put the `statement` in parentheses? The RETAIN etc. follow this, and it may end up being ambiguous at the end of a SQL query.

Another question: what is the definition of "reservoirPath"? Is it a path to a Delta table? I think allowing paths makes sense for commands like VACUUM that only operate on Delta tables, but as inputs it's not a given that the input is even a Delta table. Maybe we can leave that out and just rely on the regular:
```
delta.`/path/to/table`
```
(How do you write backticks in markdown code blocks? :) )

arunravimv replied:
I tried surrounding the query with LEFT_PAREN and RIGHT_PAREN symbols 
`statement
    : VACUUM (path=STRING | table=qualifiedName)
        (USING INVENTORY (reservoirPath=STRING | reservoirTable=qualifiedName |  LEFT_PAREN reservoirQuery=statement RIGHT_PAREN))?
        (RETAIN number HOURS)? (DRY RUN)?
`

However, I ran into EOF error for some statements that are supposed to go to delegated parser
`DELETE FROM testTable WHERE id IN (0,10,20,30,40)` with stacktrace like 

> == SQL ==
>[info] DELETE FROM testTable WHERE id IN (0,10,20,30,40)
>[info] ------------------------------------------------^^^
>[info]   at org.apache.spark.sql.catalyst.parser.ParseException.withCommand(parsers.scala:257)
>[info]   at io.delta.sql.parser.DeltaSqlParser.parse(DeltaSqlParser.scala:128)
>[info]   at io.delta.sql.parser.DeltaSqlParser.parsePlan(DeltaSqlParser.scala:77)
>[info]   at org.apache.spark.sql.SparkSession.$anonfun$sql$5(SparkSession.scala:684)
>[info]   at org.apache.spark.sql.catalyst.QueryPlanningTracker.measurePhase(QueryPlanningTracker.scala:138)
>[info]   at org.apache.spark.sql.SparkSession.$anonfun$sql$4(SparkSession.scala:683)
>[info]   at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:900)
>[info]   at org.apache.spark.sql.SparkSession.sql(SparkSession.scala:682)
>[info]   at org.apache.spark.sql.SparkSession.sql(SparkSession.scala:713)


I am still trying to see how I can fix this (If you have any pointers pls share :) ), but pushed the initial commit without the parenthesis

arunravimv replied:
regarding 
> delta.`/path/to/table`

I think if we use `reservoirTable=qualifiedName` it covers both table names and `format`. `paths` We have a logic that raises exception if the passed table is not delta. I think we can remove `reservoirPath=STRING` from the grammar and just use table or query statement alone. What do you think

bart-samwel replied:
> I think if we use reservoirTable=qualifiedName it covers both table names and format. paths We have a logic that raises exception if the passed table is not delta. I think we can remove reservoirPath=STRING from the grammar and just use table (this also covers delta.path/to/table) or a query statement alone. What do you think

+1!

bart-samwel replied:
> I am still trying to see how I can fix this (If you have any pointers pls share :) ), but pushed the initial commit without the parenthesis
You seem to have managed to fix it though, is that right?

arunravimv replied:
yes, managed to get it fixed.

bart-samwel commented on spark/src/main/scala/org/apache/spark/sql/delta/commands/VacuumCommand.scala:
```diff
@@ -125,6 +124,39 @@ object VacuumCommand extends VacuumCommandImpl with Serializable {
+        StructField("path", StringType),
+        StructField("length", LongType),
+        StructField("isDir", BooleanType),
+        StructField("modificationTime", LongType)
```
Can we put in comments to clarify the meaning? In particular:
- Is the path expected to be a URI, even for local files?
- What is the unit and epoch of modificationTime? (Unix millis or unix micros?)

bart-samwel commented on spark/src/main/scala/org/apache/spark/sql/delta/commands/VacuumCommand.scala:
```diff
@@ -125,6 +124,39 @@ object VacuumCommand extends VacuumCommandImpl with Serializable {
+        StructField("modificationTime", LongType)
+      ))
+
+    if(reservoir.schema != expectedSchema) {
```
Perhaps we could allow additional fields, i.e., just check for the fields that we need. If we want to expand the set of fields in the future then this would give users more flexibility when preparing for upgrades.

bart-samwel commented on spark/src/main/scala/org/apache/spark/sql/delta/commands/VacuumCommand.scala:
```diff
@@ -125,6 +124,39 @@ object VacuumCommand extends VacuumCommandImpl with Serializable {
+      throw new SparkException(s"reservoir schema should match ${expectedSchema.treeString}")
+    }
+
+    implicit val fileNameAndSizeEncoder
+    = org.apache.spark.sql.Encoders.product[SerializableFileStatus]
```
```suggestion
    implicit val fileNameAndSizeEncoder =
      org.apache.spark.sql.Encoders.product[SerializableFileStatus]
```

bart-samwel commented on spark/src/main/scala/org/apache/spark/sql/delta/commands/VacuumCommand.scala:
```diff
@@ -125,6 +124,39 @@ object VacuumCommand extends VacuumCommandImpl with Serializable {
+    = org.apache.spark.sql.Encoders.product[SerializableFileStatus]
+
+    reservoir
+      .filter(s"path like '$basePath/%' ")
```
Can the base path contain a `%` or `_`? That would confuse the `like`. Maybe something like:
```suggestion
      .filter(startsWith(col("path"), lit(s"$basePath/")))
```

arunravimv replied:
I am not sure if base paths can have `%`. At least on AWS, this is least likely as it falls under special characters that need extreme special handling and would often create problems in various AWS services. I think your suggestion is cleaner and I shall update the logic to use `startsWith` function.

bart-samwel commented on spark/src/main/scala/org/apache/spark/sql/delta/commands/VacuumCommand.scala:
```diff
@@ -125,6 +124,39 @@ object VacuumCommand extends VacuumCommandImpl with Serializable {
+
+    reservoir
+      .filter(s"path like '$basePath/%' ")
+      .selectExpr(s"replace(path, '$basePath/', '') as path", "length", "isDir", "modificationTime")
```
The `replace` searches the entire string, which could lead to false matches if the path is local files (not a URI) and the base path is repeated later in the path. How about using `s"substr(path, ${basePath.length+2})` or something like that?

bart-samwel reviewed (commented):
Thanks, this looks great! I put in a couple of comments and I haven't reviewed the tests yet -- I'll do that on the next round (and it's almost bedtime :) ).

arunravimv commented:
@bart-samwel I have fixed/applied all the changes/suggestions pointed out by you. Please review the changes and test cases

The only pending item is adding a parenthesis around the subquery. Somehow am running into a parse error in some pass-through conditions (eg: delete from table query using IN operator) as pointed out in the above [comment](https://github.com/delta-io/delta/pull/2257#discussion_r1379749900). The error is getting thrown from the delegated spark parser and not delta while adding parenthesis around the subquery

bart-samwel commented on spark/src/test/scala/org/apache/spark/sql/delta/DeltaVacuumSuite.scala:
```diff
@@ -224,6 +227,11 @@ trait DeltaVacuumSuiteBase extends QueryTest
         val qualified = expectedDf.map(p => fs.makeQualified(new Path(p)).toString)
         checkDatasetUnorderly(result.as[String], qualified: _*)
+      case GCByInventory(dryRun, expectedDf, retention, inventory) =>
+        Given("*** Garbage collecting Reservoir")
```
```suggestion
        Given("*** Garbage collecting using inventory")
```

bart-samwel commented on spark/src/test/scala/org/apache/spark/sql/delta/DeltaVacuumSuite.scala:
```diff
@@ -578,6 +586,230 @@ class DeltaVacuumSuite
+        CheckFiles(Seq("file2.txt", "_underscore_col_=10")),
+        LogicallyDeleteFile("_underscore_col_=10/test.txt"),
+        AdvanceClock(defaultTombstoneInterval + 1000),
+        ExpectFailure(
```
This test can probably do without all the prep work -- it will probably fail just fine on an empty table. :)

bart-samwel commented on spark/src/test/scala/org/apache/spark/sql/delta/DeltaVacuumSuite.scala:
```diff
@@ -578,6 +586,230 @@ class DeltaVacuumSuite
+        Metadata(schemaString = schema.json, partitionColumns = Seq("_underscore_col_"))
+      txn.commit(metadata :: Nil, DeltaOperations.CreateTable(metadata, isManaged = true))
+
+      val expectedInventorySchema = StructType(
```
It might be worth pulling this out into the parent class, because it's reused in a lot of tests now.

bart-samwel commented on spark/src/test/scala/org/apache/spark/sql/delta/DeltaVacuumSuite.scala:
```diff
@@ -578,6 +586,230 @@ class DeltaVacuumSuite
+        CreateFile("file2.txt", commitToActionLog = false, Map("_underscore_col_" -> "10")),
+        CreateFile("_underscore_col_=10/test.txt", true, Map("_underscore_col_" -> "10")),
+        CreateFile("_underscore_col_=10/test2.txt", false, Map("_underscore_col_" -> "10")),
+        CheckFiles(Seq("file1.txt", "_underscore_col_=10")),
+        CheckFiles(Seq("file2.txt", "_underscore_col_=10")),
```
These check for the existence of files, so we can check all of them in one go.
```suggestion
        CheckFiles(Seq(
          "file1.txt",
          "file2.txt",
          "_underscore_col_=10/test.txt",
          "_underscore_col_=10/test2.txt")),
```

bart-samwel replied:
Not seeing this one processed? I see it further down in the test but not here.

bart-samwel commented on spark/src/test/scala/org/apache/spark/sql/delta/DeltaVacuumSuite.scala:
```diff
@@ -578,6 +586,230 @@ class DeltaVacuumSuite
+    withEnvironment { (tempDir, clock) =>
+      val deltaLog = DeltaLog.forTable(spark, tempDir, clock)
+      val txn = deltaLog.startTransaction()
+      val schema = new StructType().add("_underscore_col_", IntegerType).add("n", IntegerType)
```
It's worth calling out in a comment what this is trying to do. The point is that directories that start with `_` are usually hidden files, but if they match the pattern of a partition directory, then we have to treat them as non-hidden.

bart-samwel reviewed (commented):
This is coming along nicely! I still have a couple of comments. (Sorry about the long review cycles. I have a bit too much going on at the moment. :/ )

arunravimv commented:
@bart-samwel I have fixed the function comments, adding a new delta error (pls check the state value that I have chosen) and minor improvements you suggested in unit testing while checking for files that exist. I have also rebased the code to the latest master to resolve upstream conflicts

bart-samwel commented on spark/src/main/resources/error/delta-error-classes.json:
```diff
@@ -1063,6 +1063,12 @@
+    "message" : [
+      "The schema for the specified INVENTORY does not contain all of the required fields. Required fields are: <expectedSchema>"
+    ],
+    "sqlState" : "22000"
```
I'd suggest "42000" because it's about schema (static, compilation time) while 22000 is usually about actual data (dynamic, runtime).

bart-samwel commented on spark/src/test/scala/org/apache/spark/sql/delta/DeltaVacuumSuite.scala:
```diff
@@ -578,6 +586,182 @@ class DeltaVacuumSuite
+        gcTest(deltaLog, clock)(
+          CreateFile("file1.txt", commitToActionLog = false),
+          CreateFile("file2.txt", commitToActionLog = false),
+          CreateFile("_underscore_col_=10/test.txt", false),
```
Let's call out in a comment that `_underscore_col_` is not a partition column, so then this must be treated as a hidden file.

bart-samwel commented on spark/src/test/scala/org/apache/spark/sql/delta/DeltaVacuumSuite.scala:
```diff
@@ -578,6 +586,182 @@ class DeltaVacuumSuite
+        val reservoirData = Seq(
+          Row(s"${deltaLog.dataPath}/file1.txt", 300000L, false, 0L),
+          Row(s"${deltaLog.dataPath}/file2.txt", 300000L, false, 0L),
+          Row(s"${deltaLog.dataPath}/_underscore_col_=10/test.txt", 300000L, false, 0L),
```
Same here

bart-samwel reviewed (approved):
LG, nice work! I left a couple of final suggestions but I'll approve as-is.

arunravimv commented:
@bart-samwel , thanks for guiding me through and reviewing my changes. I have fixed the final comments.

arunravimv commented:
@tdas as per our discussion today, I have rebased my changes with the latest master, please review and advise on the next steps

tdas commented:
Thank you! A bit busy with 3.1 release, I will get back to you next week. Thank you for your patience!

tdas commented:
Hey @arunravimv finally got sometime to review this PR. I am still going through it but overall looks pretty good. Will merge when I am done with the pass.

tnyz commented:
hi @tdas is there a target release version for this feature thanks

tdas commented:
@tnyz target version is next release 3.2. tentatively april (3 months since 3.1).

tdas reviewed (approved):
finished reviewing it. merging it soon.
